test/cpp/qps/client_async.cc

```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
       return nullptr;
   }
 }
+
 std::unique_ptr<Client> CreateGenericAsyncStreamingClient(
     const ClientConfig& args) {
   return std::unique_ptr<Client>(new GenericAsyncStreamingClient(args));
 }
 
+struct CallbackClientRpcData {
+  CallbackClientRpcData()
+      : thread_mu_(), thread_cv_(), rpc_done_(false), alarm_(nullptr) {}
+
+  ~CallbackClientRpcData() { delete alarm_; }
+
+  std::mutex thread_mu_;
+  std::condition_variable thread_cv_;
+  bool rpc_done_;
+  Alarm* alarm_;
+};
+
+class CallbackClient
+    : public ClientImpl<BenchmarkService::Stub, SimpleRequest> {
+ public:
+  CallbackClient(const ClientConfig& config)
+      : ClientImpl<BenchmarkService::Stub, SimpleRequest>(
+            config, BenchmarkStubCreator) {
+    num_threads_ = NumThreads(config);
+    ctxs_.resize(num_threads_);
+    threads_done_ = 0;
+    SetupLoadTest(config, num_threads_);
+  }
+
+  virtual ~CallbackClient() {}
+
+  virtual bool InitThreadFuncImpl(size_t thread_idx) { return true; }
+
+  virtual bool ThreadFuncImpl(Thread* t, size_t thread_idx) = 0;
+
+ protected:
+  size_t num_threads_;
+  std::mutex mu_;
```

Comment:
Are there some lock annotations you can add to state what this mutex is supposed to be guarding? It's currently not guarding all the variables in this structure because `ctxs_` is not protected by it. Thus it helps to be clear about where it needs to be used.

Reply:
Sure, its preventing the main thread that spawns the worker threads from ending it. I'll add comments